### PR TITLE
[Team2] Final Report

### PR DESCRIPTION
**Pull request by tomeaton17**
Added Team 2 preliminary report. Formatting for figures and tables are buggy but will fix tomorrow morning. Please review actual content of report :)

### CODE REVIEW

**Comment by tomeaton17** [comment]
I realise the references are 💥 so will fix this morning when I wake up

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 1]
```diff
@@ -124,11 +139,85 @@ \subsubsection{Design of action space}
...
+\label{survivebonus-algorithm}
+\eIf{Agent is in critical state} {
+surviveBonus $\gets$ surviveBonus $+1$\;
+}{
+surviveBonus $\gets$ surviveBonus $-(3.0\times\textrm{daysAtCriticalHealth})$\;
+}
+\If{Agent was in critical state before action}{
+surviveBonus $\gets$ surviveBonus $+(5.0\times\textrm{daysAtCriticalHealth})$\;
+}
+}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion

```
I think there is an extra '}' here preventing it from compiling correctly.

**Review by TeDand — CHANGES_REQUESTED** [review]
Also could there please be punctuation at the end of all the figure captions?

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 2]
```diff
@@ -145,3 +234,239 @@ \subsubsection{Design of reward function}
...
+5000                     & 70                                       & 0.2                            & 10                           & 797                              & 12                                        & 71.61                        \\
+5000                     & 60                                       & 0.2                            & 10                           & 830                              & 13                                        & 82.71                        \\
+5000                     & 80                                       & 0.5                            & 10                           & 859                              & 9                                         & 80.48                        \\
+5000                     & 80                                       & 1                              & 1                            & 865                              & 6                                         & 69.33                        \\ \bottomrule
+\end{tabular}
+\end{adjustbox}
+\end{table}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex:257]: I think there should be a label and caption for this table to be referenced in the text.

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 3]
```diff
@@ -124,11 +139,85 @@ \subsubsection{Design of action space}
...
+The reward function provides a way for us to set the desired behaviour of our agent. The reward function will assess the quality of a particular action under a specific state and provide feedback, either reward or punishment, to the agent. Initially, we have designed four desired behaviours. We encourage our agent to survive, not to overeat, and to save neighbour: if the neighbouring agent is in critical state, the agent will be punished. However, the last desired behaviour was only enforced by the simple communication of "Ask HP" with the agent that is one floor below. Indeed, as shown in Table \ref{cum-death}, implementing a simple communication did not improve the agent’s performance on saving other agent, i.e., the overall death count remained the same after adding the simple communication and tuned reward function accordingly. A potential reason might be that the agent cannot learn to save with such limited information. As such, in our subsequent experiments, the simple communication was not considered, and the primary focus was placed on the first two desired behaviours.
+
+\begin{table}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]: I don't know if there is an issue with this table but the horizontal line on the bottom doesn't span the entire length of the table.

**Review by TeDand — CHANGES_REQUESTED** [review]
Minor issues with tables.

File: 004_team_2_agent_design/004_team_2_agent_design.tex (hunk 2 above)
Comment (tomeaton17) [reply 004_team_2_agent_design/004_team_2_agent_design.tex:257]: Added

File: 004_team_2_agent_design/004_team_2_agent_design.tex (hunk 3 above)
Comment (tomeaton17) [reply 004_team_2_agent_design/004_team_2_agent_design.tex]: Fixed

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 4]
```diff
@@ -1,71 +1,76 @@
...
+Our goal is to create a unique advantage by allowing for the agent to experiment with different actions to different situations, until the most appropriate actions can be dynamically chosen for the future. A reward system is utilised, whereby a suitable action will be rewarded, and a detrimental action will be penalised. Given enough iterations, the agent should be sufficiently trained to consistently choose the most correct actions. This is achievable as most other agents have fixed behaviours which can be predicted by our agent; however, we also experiment with random agents to understand and quantify the change in performance and utility. Relying on a stochastic method to execute learning can result in vastly different outcomes, but it is expected that in general the agents will perform better with time.
+\subsection{In the pursuit of organisation}\label{team2-organisation}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsection{In the Pursuit of Organisation}\label{team2-organisation}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 5]
```diff
@@ -1,71 +1,76 @@
...
 
 The PHC algorithm is only able to locate an optimal policy against stationary strategy opponents and has never been proven to converge if pitted against non-stationary strategy opponents \cite{BowlingMichael2002Mlua}. As according to $Q$, that converges at some $Q'$, the mixed strategy $\pi$ is greedy and will therefore converge to the best possible policy in response to the environment the agent is placed in. 
 
-\section{Agent Design}
+\section{Agent Design and Implementation}
 \subsection{Design of states and actions}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsection{Design of States and Actions}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 6]
```diff
@@ -77,11 +82,9 @@ \subsubsection{Design of state space}
...
+
 \begin{table}
-\label{state-table}
 \centering
 \caption{State space defintion}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\caption{State space defintion.}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 7]
```diff
@@ -1,71 +1,76 @@
...
 
 The PHC algorithm is only able to locate an optimal policy against stationary strategy opponents and has never been proven to converge if pitted against non-stationary strategy opponents \cite{BowlingMichael2002Mlua}. As according to $Q$, that converges at some $Q'$, the mixed strategy $\pi$ is greedy and will therefore converge to the best possible policy in response to the environment the agent is placed in. 
 
-\section{Agent Design}
+\section{Agent Design and Implementation}
 \subsection{Design of states and actions}
 As the learning involves interaction with the environment, it is essential to define the state space and the action space. A state space is a set of states, represented by vectors of observations, that an agent could possibly experience throughout its life. An action space is a set of actions that an agent could possibly perform in the game. The design of the state space and action space is tailored to the task assigned, namely maximising both individual and collective utilities. 
 \subsubsection{Design of state space}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsubsection{Design of State Space}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 8]
```diff
@@ -104,11 +107,22 @@ \subsubsection{Design of state space}
 4399         & 90 – Max HP & 90 – Max Food    & 3                & 90 – Max HP    \\ \bottomrule
 \end{tabular}
 \end{adjustbox}
+\label{state-table}
 \end{table}
+\subsubsection{Implementation of state space}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsubsection{Implementation of State Space}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 9]
```diff
@@ -104,11 +107,22 @@ \subsubsection{Design of state space}
...
+	return a.stateSpace[hp][currFood][critDays][nHP]
+}
+\end{lstlisting}
 \subsubsection{Design of action space}
-In the game, taking food is the main interaction with the Tower, which provides feedback according to the amount of food taken. Thus, we divided the action of taking food into six distinct actions with different number of intended food intake, as shown in the table \ref{action-space}.
+In the game, taking food is the main interaction with the Tower, which provides feedback according to the amount of food taken. Thus, we divided the action of taking food into six distinct actions with different number of intended food intake, as shown in Table \ref{action-space}.
 \begin{table}
-\label{action-space}
 \centering
 \caption{Action space definition}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\caption{Action space definition.}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 10]
```diff
@@ -104,11 +107,22 @@ \subsubsection{Design of state space}
...
+\begin{lstlisting}
+func (a *CustomAgent2) CheckState() int {
+	hp := CategoriseObs(a.HP())
+	currFood := CategoriseObs(int(a.CurrPlatFood())
+	critDays := a.DaysAtCritical()
+	nHP := CategoriseObs(a.neighbourHP)
+	return a.stateSpace[hp][currFood][critDays][nHP]
+}
+\end{lstlisting}
 \subsubsection{Design of action space}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsubsection{Design of Action Space}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 11]
```diff
@@ -121,14 +135,86 @@ \subsubsection{Design of action space}
 4             & 25                   \\
 5             & 30                   \\ \bottomrule
 \end{tabular}
+\label{action-space}
 \end{table}
 
 We excluded the possibility of taking more than 30 food units, as eating above this level will only bring punishment to our agent according to our reward design. Note that the intended food intakes are multiple of 5, as we would like the agent to have less actions to learn, resulting in faster convergence. 
-
+\subsubsection{Implementation of action space}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsubsection{Implementation of Action Space}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 12]
```diff
@@ -121,14 +135,86 @@ \subsubsection{Design of action space}
...
+The action space is implemented as a 2D slice. It is iterated through and assigned the appropriate food intake according to Table \ref{action-space}.
+\begin{lstlisting}
+for i := 1; i < actionDim; i++ {
+	actionSpace[i] = actionSpace[i-1] + 5
+}
+\end{lstlisting} 
 \subsubsection{Design of reward function}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsubsection{Design of Reward Function}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 13]
```diff
@@ -121,14 +135,86 @@ \subsubsection{Design of action space}
...
+\toprule
+                                              &                                                                                   & Without communication & \begin{tabular}[c]{@{}l@{}}With simple\\ communication\end{tabular} \\ \midrule
+\multirow{2}{*}{Cumulative death in 500 days} & Homogeneous tower (14 team 2 agent)                                               & 50                    & 78                                                                  \\
+                                              & \begin{tabular}[c]{@{}l@{}}Heterogeneous\\ Tower (2 agents per team)\end{tabular} & 133                   & 139                                                                 \\ \bottomrule
+\end{tabular}
+\caption{Effect of simple communication on cumulative death count in 500 days.}
+\label{cum-death}
+\end{table}
+
+\subsubsection{Implementation of reward function}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsubsection{Implementation of Reward Function}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 14]
```diff
@@ -121,14 +135,86 @@ \subsubsection{Design of action space}
...
+\caption{Saving bonus algorithm}
+\label{savingbonus-algorithm}
+\eIf{Agent on floor below is in critical state} {
+savingBonus $\gets$ savingBonus $-3$\;
+}{
+savingBonus $\gets$ savingBonus $+1$\;
+}
+\end{algorithm}
+
+\subsection{Simple reward function}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsection{Simple Reward Function}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 15]
```diff
@@ -145,3 +231,241 @@ \subsubsection{Design of reward function}
...
+	\item Tuning the overeat punishment parameter seemed to have little impact on agent behaviour as the agent would understand punishment regardless of severity and try to stay below the overeat threshold. An optimal value was found at 0.2*(food taken that day)
+	\item For tuning the critical health status punishment parameter, if agent was punished less for having consecutively critical hp meant fewer overall deaths, but more team2 deaths. Setting a stronger punishment for this gave the opposite effect. The value for this parameter giving best performance on average was at 10*(days agent has stayed at critical consecutively)
+\end{enumerate}
+
+The values found for tuning are only estimates of what gives best agent performance in synergy with the tower. This is because the tower introduces a very strong randomness effect as it reshuffles the agent floor locations after only one day. This means almost no run will ever be the same and highly unlucky events, with for example no food for several days, are very likely to occur. 
+
+\section{Results and analysis}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\section{Results and Analysis}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 16]
```diff
@@ -145,3 +231,241 @@ \subsubsection{Design of reward function}
...
+\subsection{Aim of the simulations}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsection{Aim of the Simulations}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 17]
```diff
@@ -145,3 +231,241 @@ \subsubsection{Design of reward function}
...
+
+The performance of our agent will be justified by metric relating to both individual and collective utility: reward by day, cumulative reward by day, HP vs day, and food taken per day (HP), and age lived per generation.
+
+\subsection{The performance of our agent}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsection{The Performance of our Agent}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 18]
```diff
@@ -145,3 +231,241 @@ \subsubsection{Design of reward function}
...
+\label{avhp-team2}
+\end{figure}
+
+The results in Figure \ref{avhp-team2} are from an experiment without random agents. Only one agent of our team is included, and reincarnation is allowed. 
+
+Following the design in the reward function, the HP of our agent is learnt to retain under 80. It does not choose to stay as close as possible to the upper HP limit 80. Instead, the agent decides to maintain the HP around 55 on average. 
+
+This may indicate that staying at a high HP level may not be a good surviving strategy in the tower. 
+
+\subsection{Reincarnation of agent}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsection{Reincarnation of Agent}
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 19]
```diff
@@ -145,3 +231,241 @@ \subsubsection{Design of reward function}
...
+\end{figure}
+
+\begin{figure}
+\centering
+\includegraphics{004_team_2_agent_design/252team2}
+\caption{Tower of only Team 2 agents reaches a point where all agents stays alive.}
+\label{252team2}
+\end{figure}
+
+\subsection{Learnt policies}
```
Comment (TeDand) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
\subsection{Learnt Policies}
```

**Review by TeDand — DISMISSED** [review]
Just a few final comments but looks good to me :)

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 20]
```diff
@@ -1,71 +1,76 @@
...
+The objective of this agent design was to implement a form of reinforcement learning to allow for an increase in performance and thus utility over time. This must be met without compromising on the overall goal of the simulations, whereby the agent must attempt to strike a balance between prioritising the micro-level goal of maximal individual utility or prioritising the macro-level goal of maximising collective utility through sustainable resource management, the latter involving a level of cooperation with the other agent types. Certain parallels can be drawn to evolutionary ecology, as the Tower with the Resident agents can represent a closed ecosystem, with the Platform providing a limited food source.
```
Comment (daryllimyt) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]: Should 'platform' have its first letter capitalized?

Comment (daryllimyt) [reply 004_team_2_agent_design/004_team_2_agent_design.tex]: Same sentiment for 'tower'

Comment (daryllimyt) [reply 004_team_2_agent_design/004_team_2_agent_design.tex]: We should at least standardize this across the report

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 21]
```diff
@@ -121,14 +135,86 @@ \subsubsection{Design of action space}
...
 
-Based on the health decay model defined in \ref{health-modelling}, the agent reward was designed to encourage the agent to reward a health above critical and punish dropping to critical health as well as punish overeating using a reward point value for each day. The reward function with its tuned hyperparameters is as such:
+\begin{table}[]
+\begin{tabular}{@{}llll@{}}
+\toprule
+                                              &                                                                                   & Without communication & \begin{tabular}[c]{@{}l@{}}With simple\\ communication\end{tabular} \\ \midrule
+\multirow{2}{*}{Cumulative death in 500 days} & Homogeneous tower (14 team 2 agent)                                               & 50                    & 78                                                                  \\
+                                              & \begin{tabular}[c]{@{}l@{}}Heterogeneous\\ Tower (2 agents per team)\end{tabular} & 133                   & 139                                                                 \\ \bottomrule
+\end{tabular}
```
Comment (daryllimyt) [inline 004_team_2_agent_design/004_team_2_agent_design.tex:160]: This table overflows

**Review by daryllimyt — DISMISSED** [review]

File: 004_team_2_agent_design/004_team_2_agent_design.tex (hunk 20 above)
Comment (tomeaton17) [reply 004_team_2_agent_design/004_team_2_agent_design.tex]: Yes I will not capitalise.

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 22]
```diff
@@ -145,3 +233,241 @@ \subsubsection{Design of reward function}
...
+HP     & Food decided to take \\ \midrule
+30-100 & 5.0                  \\ \bottomrule
+\end{tabular}
+\end{table}
+
+Table \ref{30100-team2} shows that when the HP of agent is larger than 30, it tends to eat the minimum amount of food and leave the food to others.
+
+In summary, from the observations gathered, the learnt policy of the agent follows certain patterns:
+\begin{itemize}
+	\item The priority is the survival of itself; the tower scenario is often too unpredictable to switch to more altruistic objectives
```
Comment (mdubied) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
	\item The priority is the survival of itself; the tower scenario is often too unpredictable to switch to more altruistic objectives.
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 23]
```diff
@@ -145,3 +233,241 @@ \subsubsection{Design of reward function}
...
+	\item Leaving food for others when the food is not sufficient tends to bring more collective benefit in the long term
```
Comment (mdubied) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
	\item Leaving food for others when the food is not sufficient tends to bring more collective benefit in the long term.
```

File: 004_team_2_agent_design/004_team_2_agent_design.tex [hunk 24]
```diff
@@ -145,3 +233,241 @@ \subsubsection{Design of reward function}
...
+	\item Maintaining HP over 30 is safer for the survival of agent and is more likely to result in a Pareto optimal value for individual utility
```
Comment (mdubied) [inline 004_team_2_agent_design/004_team_2_agent_design.tex]:
```suggestion
	\item Maintaining HP over 30 is safer for the survival of agent and is more likely to result in a Pareto optimal value for individual utility.
```

**Review by mdubied — CHANGES_REQUESTED** [review]
Looks good! Can you just change the caption of your tables to be all _below_ the tables?

**Review by mdubied — DISMISSED** [review]
LGTM. Great job guys!

**Review by daryllimyt — DISMISSED** [review]
LGTM

**Review by daryllimyt — APPROVED** [review]

**Review by mdubied — APPROVED** [review]
Re-approve

**Pull request merged**